src/main/java/org/mate/bookstorespringboot/service/impl/ShoppingCartIServiceImpl.java

```diff
@@ -0,0 +1,117 @@
+package org.mate.bookstorespringboot.service.impl;
+
+import lombok.RequiredArgsConstructor;
+import org.mate.bookstorespringboot.controller.dto.cartitem.CartItemRequestDto;
+import org.mate.bookstorespringboot.controller.dto.cartitem.UpdateCartItemRequestDto;
+import org.mate.bookstorespringboot.controller.dto.shoppingcart.ShoppingCartDto;
+import org.mate.bookstorespringboot.controller.mapper.CartItemMapper;
+import org.mate.bookstorespringboot.controller.mapper.ShoppingCartMapper;
+import org.mate.bookstorespringboot.exceptions.EntityNotFoundException;
+import org.mate.bookstorespringboot.model.Book;
+import org.mate.bookstorespringboot.model.CartItem;
+import org.mate.bookstorespringboot.model.ShoppingCart;
+import org.mate.bookstorespringboot.model.User;
+import org.mate.bookstorespringboot.repository.BookRepository;
+import org.mate.bookstorespringboot.repository.CartItemRepository;
+import org.mate.bookstorespringboot.repository.ShoppingCartRepository;
+import org.mate.bookstorespringboot.service.interfaces.ShoppingCartService;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartIServiceImpl implements ShoppingCartService {
+
+    private static final String SHOPPING_CART_NOT_FOUND = "Shopping cart not found with id: ";
+    private static final String CART_ITEM_NOT_FOUND = "Cart item not found with id: ";
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final CartItemRepository cartItemRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final BookRepository bookRepository;
+    private final CartItemMapper cartItemMapper;
+
+    @Override
+    public ShoppingCartDto getShoppingCart(Long userId) {
+        ShoppingCart cart = getShoppingCartByUserId(userId);
+        return shoppingCartMapper.toDto(cart);
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCartDto addCartItemToShoppingCart(Long userId,
+                                                     CartItemRequestDto cartItem) {
+        ShoppingCart shoppingCartToBeSaved = getShoppingCartByUserId(userId);
+
+        Book book = bookRepository.findById(cartItem.bookId())
+                .orElseThrow(() -> new EntityNotFoundException("Book not found with id: "
+                        + cartItem.bookId()));
+
+        cartItemRepository.findByShoppingCartIdAndBookId(shoppingCartToBeSaved.getId(),
+                        cartItem.bookId())
+                .ifPresent(cartItem1 -> {
+                    throw new EntityNotFoundException("Book with id: "
+                            + cartItem.bookId()
+                            + " already in shopping cart");
+                });
```

Comment:
Do you really need to throw an exception in this case? Maybe increase CartItem quantity?